src/backends/policy_engine/souffle/datalog_lowering_visitor_test.cc

```diff
@@ -0,0 +1,243 @@
+     .datalog_is_operation = &kStringLiteralDatalogIsOperation},
+    {.ir_operation = &kMergeOpIrOperation,
+     .datalog_is_operation = &kMergeOpIsOperation}};
+
+INSTANTIATE_TEST_SUITE_P(IrOperationLoweringTest, IrOperationLoweringTest,
+                         ValuesIn(kIrAndDatalogOperations));
+
+struct ModuleAndExpectedRakshaDatalogFactBuilders {
+  std::function<ir::Module()> module_fn;
+  std::function<RakshaDatalogFacts()> facts_fn;
+};
+
+class DatalogLoweringVisitModuleTest
+    : public TestWithParam<ModuleAndExpectedRakshaDatalogFactBuilders> {};
+
+TEST_P(DatalogLoweringVisitModuleTest, DatalogLoweringVisitModuleTest) {
+  const auto [module_fn, facts_fn] = GetParam();
+  ir::Module module = module_fn();
+  RakshaDatalogFacts raksha_datalog_facts = facts_fn();
+  ir::SsaNames ssa_names;
+  DatalogLoweringVisitor visitor(ssa_names);
+  module.Accept(visitor);
+  EXPECT_EQ(visitor.datalog_facts().ToDatalogString(),
+            raksha_datalog_facts.ToDatalogString());
+}
+
+static const ModuleAndExpectedRakshaDatalogFactBuilders
+    kModuleAndExpectedRakshaDatalogFacts[] = {
+        // Empty module test.
+        {.module_fn = []() { return ir::Module(); },
+         .facts_fn = []() { return RakshaDatalogFacts(); }},
+        // Test containing a single passthrough operation.
+        {.module_fn =
+             []() {
+               ir::BlockBuilder block_builder;
+               block_builder.AddOperation(kMergeOpOperator, {},
+                                          {ir::Value(ir::value::Any())});
+               ir::Module module;
+               module.AddBlock(block_builder.build());
+               return module;
+             },
+         .facts_fn =
+             []() {
+               RakshaDatalogFacts facts;
+               facts.AddIsOperationFact(
+                   ir::datalog::IsOperationFact(ir::datalog::Operation(
+                       ir::datalog::Symbol("UNKNOWN"),
+                       ir::datalog::Symbol("sql.MergeOp"),
+                       ir::datalog::Symbol("%0.out"),
+                       ir::datalog::OperandList(ir::datalog::Symbol("<<ANY>>"),
+                                                ir::datalog::OperandList()),
+                       ir::datalog::AttributeList())));
+               return facts;
+             }},
+        // Show two passthroughs chained together.
+        {.module_fn =
+             []() {
+               ir::BlockBuilder block_builder;
+               const ir::Operation &operation = block_builder.AddOperation(
+                   kMergeOpOperator, {}, {ir::Value(ir::value::Any())});
+               block_builder.AddOperation(
+                   kMergeOpOperator, {},
+                   {ir::Value(ir::value::OperationResult(operation, "out"))});
+               ir::Module module;
+               module.AddBlock(block_builder.build());
+               return module;
+             },
+         .facts_fn =
+             []() {
```

Comment:
Similarly, can we just have a list of `std::vector<InputRelationFact>`instead of lambdas?

Reply:
That wouldn't be enough to solve our problem. The main issue here is that test parameter values must be copyable. `IsOperation`, the main `InputRelationFact` I am currently concerned with, has `Record` members, which contain `std::unique_ptr`s, which are not copyable.

Reply:
Perhaps, it is time to change `unique_ptr` to `intrusive_ptr`? (Not in this PR, but as a follow up.) Nevertheless, is there a way to make this test more readable?

Reply:
I toyed with this for a while. I think it's possible, but I'm not sure that it's a good idea.

Semantically, we want `std::unique_ptr`. These Souffle mirror structures are the owners of their child objects and they should be the only objects holding a strong pointer to those objects. The friction that we are feeling around the use of these objects is entirely the result of non-copyable objects just being tricky to work with.

It is also notable that this friction isn't really a problem in our production code, only in our test code. While I am quite willing to make changes to production code to make it more testable, it feels wrong to make the ownership model more complex (ie, moving from one owner to many owners) just to make tests easier to write. I feel that we should continue to use `std::unique_ptr` for these Souffle mirror objects.

Reply:
+1, we should not change production code just for convenience of tests.